---
title: "assembly_stats"
author: "Allan Zuza"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_packages}
library(janitor)
library(kableExtra)
library(tidyverse)
library(gridExtra)

```


## import assembly statistics
```{r}
library()

assembly_headers <- read.table("results/assembly-stats/colnames_100478.tsv",
           header = TRUE)

assembly_headers

assembly_stats <- read.table("results/assembly-stats/100478.tsv",
           header = FALSE)

assembly_stats

colnames(assembly_stats) <- c(colnames(assembly_headers), "assembler")
assembly_stats
```

## summarise the stats
```{r, message=FALSE}
assembly_stats %>%
  group_by(assembler) %>%
  summarise(number_of_contigs = mean(number), total_length = mean(total_length), longest = mean(longest), shortest = mean(shortest), n50 = mean(N50))

assembly_stats %>%
  group_by(assembler) %>%
  ggplot() +
  geom_point(aes(assembler, log10(total_length)), color = "red") + ylim(0,7.5) +
  geom_point(aes(assembler, log10(number)), color = "blue") +
  geom_point(aes(assembler, log10(shortest)), color = "purple")


```

# comparing the hybrid assemblies
the assemblies have differing numbers and types of plasmids. using pubmlst online
platform the plamids are distributed as below in the hybrid assemblies
```{r}

final_assembly_stats <- read_csv("docs/assembly_statistics.csv")

final_assembly_stats
summary(final_assembly_stats)

column_names <- final_assembly_stats %>%
  names() %>%
  str_replace_all("_", " ") %>%
  str_to_title()

table_1 <- final_assembly_stats %>%
  knitr::kable(col.names =
                 linebreak(column_names, align = "l"),
               caption = "Assembly Statistics Of Hybrid Assemblies") %>%
  kable_styling(latex_options = c("striped", "hover", "condensed", "responsive")) %>%
  column_spec(1, bold = TRUE) %>%
  add_header_above(header = c(" " = 4, "Plasmid Replicons" = 9 ),
                   font_size = 17) %>%
  kable_classic_2(full_width = FALSE) %>%
  row_spec(0, bold = TRUE) %>%
  footnote(general = c("* Not an exact match"))

table_1
```


#save table 1
```{r}
outdir <- "results/plots/r"
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "Table1.html")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

save_kable(table_1, file = plot_name,
           self_contained = TRUE)

```


## mauve comparisons
the assemblies look quite similar in mauve. plasmids are responsible for most of the variation seen

```{r}

#knitr::include_graphics("results/plots/r/2024-07-09/distribution_of_isolates.pdf")
getwd()
```

## compare the genomes
### load the libraries
```{r}
library(genoPlotR)
```


### load the genomes and comparisons
```{r eval=FALSE}

dna_seg1 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/1004078.gbk")
dna_seg2 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/BKREGE.gbk")
dna_seg3 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/CAA8N9.gbk")
dna_seg4 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/CAAUE8.gbk")
dna_seg5 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/CHI11E.gbk")

save(dna_seg1, dna_seg2, dna_seg3, dna_seg4, dna_seg5,
     file = "data/r/dnaSegments.RData")

comparison_1 <- read_comparison_from_blast("docs/blast_comparisons/1004078-BKREGE.txt")
comparison_2 <- read_comparison_from_blast("docs/blast_comparisons/BKREGE-CAA8N9.txt")
comparison_3 <- read_comparison_from_blast("docs/blast_comparisons/CAA8N9-CAAUE8.txt")
comparison_4 <- read_comparison_from_blast("docs/blast_comparisons/CAAUE8-CHI11E.txt")
```


```{r}
load(file = "data/r/dnaSegments.RData")

region_num <- c("one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen")
starts <- c("cpxR", "ahr", "pknK", "ppiB", "phoH", "ves", "dcuS", "lcfB", "budC", "ychF", "yedJ", "fadL", "nadK", "cysD", "tsaD", "pagL")
ends <- c("fieF" ,"xynB", "proB", "cadC", "gpx1", "nifJ", "ynfM", "yjhB", "ccmF", "ais", "ddpB", "gltX", "bepF", "casC", "hdeD", "ycnE" )

regions <- tibble(region_num, starts, ends)

regions

```


### draw the segments
```{r functions_for_the_regions}

# a function for creating the segments for specific regions
create_segment <- function(segment, x, y){
  positions <- c(which(segment$name == x), which(segment$name == y))
  positions <- sort(positions)
  segment[positions[1]:positions[2], ]
}

# a function for creating annotation files for the region created above
create_annotation <- function(segment) {
    annotations = 
      annotation(text = segment$gene,
                         x1 = middle(segment),
                         x2 = middle(segment),
                         rot = 30)
}


# function to plot the region in all the assemblies
plot_region <- 
  function(region, title) {
  main = title
    
  x <- regions[regions$region_num == region, ]$starts
  y <- regions[regions$region_num == region, ]$ends
  
  seg1 <- create_segment(dna_seg1, x, y)
  seg1_annotation <- create_annotation(dna_seg1)
  
  seg2 <- create_segment(dna_seg2, x, y)
  seg2_annotation <- create_annotation(dna_seg2)
  
  seg3 <- create_segment(dna_seg3, x, y)
  seg3_annotation <- create_annotation(dna_seg3)
  
  seg4 <- create_segment(dna_seg4, x, y)
  seg4_annotation <- create_annotation(dna_seg4)
  
  seg5 <- create_segment(dna_seg5, x, y)
  seg5_annotation <- create_annotation(dna_seg5)
  
  
  plot_gene_map(list(seg1, seg2, seg3, seg4, seg5),
              gene_type = "arrows",
              annotations = list(seg1_annotation, seg2_annotation,
                                 seg3_annotation, seg4_annotation,
                                 seg5_annotation),
              annotation_height = 2,
              comparisons = list(comparison_1, comparison_2, comparison_3, comparison_4),
              dna_seg_labels = list("1004078", "BKREGE",
                                    "CAA8N9", "CAAUE8", "CHI11E"),
              main = main,
              dna_seg_scale = TRUE)
  
  }


# plotting the regions
plot_region("two", " Region 2")

plot_region("one", "Gap 17")


plot_region("two") + plot_region("two") + plot_region("two")
```

