---
title: "Assessing Hybrid Assemblies"
author: "Allan Zuza"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_packages}
# install.packages("pacman")

pacman::p_load(janitor,
               kableExtra,
               tidyverse,
               gridExtra,
               genoPlotR,
               patchwork)

```


# import assembly statistics
```{r}

assembly_headers <- read.table("results/assembly-stats/colnames_100478.tsv",
           header = TRUE)

assembly_headers

assembly_stats <- read.table("results/assembly-stats/100478.tsv",
           header = FALSE)
# add column names from the assembly header df
colnames(assembly_stats) <- c(colnames(assembly_headers), "assembler")

head(assembly_stats)
```

## summarise the stats
```{r, message=FALSE, eval=FALSE, echo=FALSE}
assembly_stats %>%
  group_by(assembler) %>%
  summarise(number_of_contigs = mean(number), total_length = mean(total_length), longest = mean(longest), shortest = mean(shortest), n50 = mean(N50))

assembly_stats %>%
  group_by(assembler) %>%
  ggplot() +
  geom_point(aes(assembler, log10(total_length)), color = "red") + ylim(0,7.5) +
  geom_point(aes(assembler, log10(number)), color = "blue") +
  geom_point(aes(assembler, log10(shortest)), color = "purple")


```

## comparing the hybrid assemblies
the assemblies have differing numbers and types of plasmids. using pubmlst online
platform the plamids are distributed as below in the hybrid assemblies
```{r}
# information of the hybrid genomes
final_assembly_stats <- read_csv("docs/assembly_statistics.csv")

plasmidfinder <- read_table("./docs/hybrid_assembly_plasmid_finder.txt") #%>%
  #select(!c("NUM_FOUND"))

# plasmid data
plasmids <-
  plasmidfinder %>%
  as.data.frame()


rownames(plasmids) <-
  plasmids$`#FILE`

plasmid_data <-
  select(plasmids, !c(`#FILE`, NUM_FOUND))

# replace all numeric to binary by matching the missing values
for (i in colnames(plasmid_data)) {
  plasmid_data[ ,i] <- if_else(plasmid_data[ ,i] %in% ".", "No", "Yes")
}


plasmid_data_1 <-
  plasmid_data %>%
  mutate(across(where(is.numeric), as.factor))

plasmid_data_1$strain <- rownames(plasmid_data_1)

# combine contig info and plasmidfinder data
table <-
  merge(final_assembly_stats, plasmid_data_1) %>%
  as_tibble() %>% rename("Strain" = "strain")

# add superscripts to the table 
x <- table %>% as.data.frame()
colnames(x)[3] <- paste0("Total Assembly Length", footnote_marker_symbol(1))
colnames(x)[4] <- paste0("Chromosome Length", footnote_marker_symbol(1))

# format the table
knitr::kable(x, escape = FALSE) %>%
  footnote(symbol = "Values in nucleotide basepairs") %>%
  kable_styling(latex_options = c("striped", "hover", "condensed", "responsive")) %>%
  column_spec(1, bold = TRUE) %>%
  add_header_above(header = c(" " = 4, "Plasmid Replicons" = 6 ),
                   font_size = 17) #%>%
  kable_classic_2(full_width = FALSE)
  
```


## save Table 1 to file
```{r, eval=FALSE}
outdir <- "results/plots/r"
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "Table1.html")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

save_kable(table_1, file = plot_name,
           self_contained = TRUE)

```


# Genome comparisons and plotting using genoPlotR
the assemblies look quite similar in mauve. plasmids are responsible for most of the variation seen


## load the annotated genbank and blast comparison files
```{r eval=FALSE}
# load prokka annotated genbank files
dna_seg1 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/1004078.gbk")
dna_seg2 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/BKREGE.gbk")
dna_seg3 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/CAA8N9.gbk")
dna_seg4 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/CAAUE8.gbk")
dna_seg5 <- read_dna_seg_from_genbank(file = "results/mauve/fasta/prokka_chromosome/CHI11E.gbk")

# save objects to Rdata
save(dna_seg1, dna_seg2, dna_seg3, dna_seg4, dna_seg5,
     file = "data/r/dnaSegments.RData")


# load files with manually curated annotations
curated_dna_seg1 <- read_dna_seg_from_genbank(file = "results/gbk/chromosome/1004078.gbk")
curated_dna_seg2 <- read_dna_seg_from_genbank(file = "results/gbk/chromosome/BKREGE.gbk")
curated_dna_seg3 <- read_dna_seg_from_genbank(file = "results/gbk/chromosome/CAA8N9.gbk")
curated_dna_seg4 <- read_dna_seg_from_genbank(file = "results/gbk/chromosome/CAAUE8.gbk")
curated_dna_seg5 <- read_dna_seg_from_genbank(file = "results/gbk/chromosome/CHI11E.gbk")

# save objects to Rdata
save(curated_dna_seg1, curated_dna_seg2, curated_dna_seg3, curated_dna_seg4, curated_dna_seg5,
     file = "data/r/curatedDnaSegments.RData")

#load comparison files created by blast
comparison_1 <- read_comparison_from_blast("docs/blast_comparisons/1004078-BKREGE.txt")
comparison_2 <- read_comparison_from_blast("docs/blast_comparisons/BKREGE-CAA8N9.txt")
comparison_3 <- read_comparison_from_blast("docs/blast_comparisons/CAA8N9-CAAUE8.txt")
comparison_4 <- read_comparison_from_blast("docs/blast_comparisons/CAAUE8-CHI11E.txt")

# save comparisons to file
save(comparison_1, comparison_2, comparison_3, comparison_4,
     file = "data/r/comparisonFiles.RData")
```


## regions with gaps in assemblies

```{r}
# region names
region_num <- c("insertion_4", "insertion_3", "insertion_2", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "phage_region_1", "sixteen", "insertion_1", "phage_region_2", "phage_region_3", "phage_region_4")

# start gene in annotation
starts <- c("mug", "mdoC", "yedJ", "cpxR", "ahr", "pknK", "ppiB", "phoH", "ves", "dcuS", "lcfB", "budC", "ychF", "yedJ", "fadL", "nadK", "cysD", "ureG", "pagL", "smpB", "sapB", "hiuH", "menA")

# end gene in annotation
ends <- c("evgA", "phoH", "yeeO", "fieF" ,"xynB", "proB", "cadC", "gpx1", "nifJ", "ynfM", "yjhB", "ccmF", "ais", "ddpB", "gltX", "bepF", "casC", "glpF_2", "ycnE", "bepF", "ribA", "folD", "cpxA" )

regions <- tibble(region_num, starts, ends)

regions

```


## create functions to use for dna segment visualisation
```{r functions_for_the_regions}
load(file = "data/r/curatedDnaSegments.RData")
load(file = "data/r/comparisonFiles.RData")


# a function for creating the segments for specific regions
create_segment <- function(segment, x, y){
  positions <- c(which(segment$name == x), which(segment$name == y))
  positions <- sort(positions)
  segment[positions[1]:positions[2], ]
}

# a function for creating annotation files for the region created above
create_annotation <- function(segment) {
    annotations = 
      annotation(text = segment$gene,
                         x1 = middle(segment),
                         x2 = middle(segment),
                         rot = 30)
}

# function to extract the dna segment data frame
get_seg <-
  function(region) {
    
  x <- regions[regions$region_num == region, ]$starts
  y <- regions[regions$region_num == region, ]$ends
  
  seg <- create_segment(dna_seg2, x, y)
  
  return(seg)
  }

# function to plot the region in all the assemblies
plot_region <- 
  function(region, title) { # takes in the region to plot and the title for the plot
  main = title
    
  x <- regions[regions$region_num == region, ]$starts
  y <- regions[regions$region_num == region, ]$ends
  
  seg1 <- create_segment(dna_seg1, x, y)
  seg1_annotation <- create_annotation(dna_seg1)

  seg2 <- create_segment(dna_seg2, x, y)
  seg2_annotation <- create_annotation(dna_seg2)
  
  seg3 <- create_segment(dna_seg3, x, y)
  seg3_annotation <- create_annotation(dna_seg3)
  
  seg4 <- create_segment(dna_seg4, x, y)
  seg4_annotation <- create_annotation(dna_seg4)
  
  seg5 <- create_segment(dna_seg5, x, y)
  seg5_annotation <- create_annotation(dna_seg5)
  
  
  plot_gene_map(list(seg1, seg2, seg3, seg4, seg5),
              gene_type = "arrows",
              annotations = list(seg1_annotation, seg2_annotation,
                                 seg3_annotation, seg4_annotation,
                                 seg5_annotation),
              annotation_height = 2,
              comparisons = list(comparison_1, comparison_2,
                                 comparison_3, comparison_4),
              dna_seg_labels = list("1004078", "BKREGE",
                                    "CAA8N9", "CAAUE8", "CHI11E"),
              main = main,
              dna_seg_scale = TRUE)
  
  }

```


## plotting regions of the chromosome with rearrangements
### The nodes with ~100 isolates as observed from Graphia

```{r plot_the_dna_segments}

plot_region("insertion_2", "Region Two")

```
This region has 144 node (might be in 144 isolates). the region shows a gain in
CDSs in the 2017 isolartes and an immediate loss of the region in the
2019 iisolates. the isolate from october 2019 (CAAUE8) gained some additional
CDSs but the structure was slighly diffeent from that of the 2017 isolates.
the isolate from december 2019 had lost most of this region and maintained a region
smaller than the one in the 2010 isolate

```{r}

# the insertion into outbreak isolates
plot_region("insertion_1", "Region One")


```

The region has a rep gene next to an intA, recF and xerC genes.
This is not typical behaviour as these genes have very different functions.
Genes in bacteria usually cluster into regions of similar function[salmonella paper].
Having genes responsible for controling DNA repair close to the rep gen which
is used to control plasmid DNA replication might have a suynergy by
allowing centralised control of these genes in the outbreak isolates. 

```{r}

# the insertion into outbreak isolates
plot_region("insertion_3", "Region Three")


```


```{r}

# the insertion into outbreak isolates
plot_region("insertion_4", "Region_4")

```
this region has been reducing in size from. this region has lost te traR gene which
is crutial for regulating conjugative plasmid transfer. this region might also
have a prophage since it has the pgFl gene which encodes a prophage tail shealth gene


```{r, echo=FALSE, eval=FALSE}

insertion1 <- get_seg("insertion_1")
insertion1$region <- "One"

insertion2 <- get_seg("insertion_2")
insertion2$region <- "Two"

insertion3 <- get_seg("insertion_3")
insertion3$region <- "Three"

insertion4 <- get_seg("insertion_4")
insertion4$region <- "Four"

insertions <-
  rbind(insertion1, insertion2, insertion3, insertion4) 

# remove hypothetical genes and save to Rdata
cds <- insertions %>%
  filter(gene != "-")
save(cds, file = "data/r/outbreakCDSs.RData")
```


### Create Phage region annotations
```{r Phage_annotation}
# these plots have the vPathPath type of obect and I have not found a way of 
# combining them into a grid using R
plot_region("phage_region_1", "Phage region 1")

plot_region("phage_region_2", "Phage region 2")

plot_region("phage_region_3", "Phage region 3")

plot_region("phage_region_4", "Phage region 4")
  
```
### repeate the phage drawing. but this time use the manually curated annotations
Prokka anotations miss insertion elements and do not prperly annotate AMR genes.
To fix this I use ncbi serches of the nucleotide sequences to get proper annotations
for the coding sequences

## Chromosomap amr insertion region

```{r chromosmal_insertion_region}

amr <- c("tet(A)", "Sul1",  "dfrA7", "CatA1")
transposases <- c( "IS1R", "ISEcl3",  "ISPrst2", "ISPrst3", "IS1R",
                   "TnAs3", "IS26", "tetR")
mercury_resistance <-  c("merT", "merP", "merC", "merA")


outbreak_insertion <-
  curated_dna_seg3 %>% filter(start >= 3359021 & end <= 3411255)

outbreak_insertion

annotate_seg <-
  function(dna_seg) {
    dna_seg$fill <-
        case_when(dna_seg$gene %in% mercury_resistance ~ "#1f78b4",
            TRUE ~ dna_seg$fill)
    
    dna_seg$fill <-
        case_when(dna_seg$gene %in% amr ~ "#a6cee3",
            TRUE ~ dna_seg$fill)
    
     dna_seg$fill <-
        case_when(dna_seg$gene %in% transposases ~ "#b2df8a",
            TRUE ~ dna_seg$fill)
     
     dna_seg$fill <-
        case_when(dna_seg$gene %in% "-" ~ "gray",
            TRUE ~ dna_seg$fill)
     
    #color the segments
    dna_seg$col <-
        case_when(dna_seg$gene %in% mercury_resistance ~ "#1f78b4",
            TRUE ~ dna_seg$fill)
    
    dna_seg$col <-
        case_when(dna_seg$gene %in% amr ~ "#a6cee3",
            TRUE ~ dna_seg$fill)
    
     dna_seg$col <-
        case_when(dna_seg$gene %in% transposases ~ "#b2df8a",
            TRUE ~ dna_seg$fill)
     
     dna_seg$col <-
        case_when(dna_seg$gene %in% "-" ~ "gray",
            TRUE ~ dna_seg$fill)
     
     return(dna_seg)

  }



outbreak_segment <-
  curated_dna_seg3 %>% filter(start >= 3359021 & end <= 3639516)


outbreak <- annotate_seg(outbreak_segment)

plot_gene_map(dna_segs = list(outbreak))
#plot the gene map

plot_gene_map(dna_segs = list(chromsomal_amr_integration),
                              gene_type = "arrows",
              annotations = annotation(text = chromsomal_amr_integration$gene,
                                          x1 = middle(chromsomal_amr_integration),
                                       rot = 45),
                                       annotation_height = 4,
              annotation_cex = 1,
              main = "chromosomal integration site in strain CHI11E",
              dna_seg_scale = TRUE)
```


