---
title: "assembly_stats"
author: "Allan Zuza"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy()
```

```{r setup, include=FALSE, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# load packages and set functions
```{r load_packages, message=FALSE, warning=FALSE}
# Load the janitor package for data cleaning
library(janitor)
# Load the kableExtra package for creating tables
library(kableExtra)
# Load the tidyverse package for data manipulation
library(tidyverse)
# Load the reshape2 package for reshaping data
library(reshape2)
# Load the ggtree package for creating tree plots
library(ggtree) 
# Load the phangorn package for phylogenetic analysis
library(phangorn)
# Load the ggnewscale package for scaling plots
library(ggnewscale)
# Load the patchwork package for combining plots
library(patchwork)
# Load the rPinecone package for working with Pinecone
library(rPinecone)
# Load the gridExtra package for arranging plots
library(gridExtra)
# Load the khroma package for managing colors
library(khroma) 
# Load the adegenet package for converting MSA to DNAbin
library(adegenet) 
# Define a custom operator for negation
'%notin%' <- Negate('%in%')
```

# plot attributes
```{r set_plot_atributes}
# set plot colors

# ward colors
ward_col <- c("CHATINKHA NURSERY" = "#CC6677",
              "PAEDIATRIC SPECIAL CARE WARD" = "#332288",
              "PAEDIATRIC MOYO WARD" = "#ddcc77",
              "4A FEMALE MEDICAL WARD" = "#117733",
              "3B MALE MEDICAL WARD" = "#88ccee",
              "PAEDIATRIC ONCOLOGY" = "#882255",
              "INTENSIVE CARE UNIT" = "#44aa99",
              "PAEDIATRIC A&E" = "#999933",
              "PAEDIATRIC SURGICAL WARD" = "#aa4499",
               "PICU/HDU" = "#7570b3",
              "unknown" = "#555555",
              "Other wards" = "#555555")
  

datecol <- c("Other" = "#f0f0f0", "2016" = "#cccccc",
             "2017" = "#969696", "2018" = "#636363",
             "2019" = "#252525" )

one_col <- c("0" = "#f0f0f0", "1" = "#117733")

two_col <- c("0" = "#f0f0f0", "1" = "#7570b3")

three_col <- c("0" = "#f0f0f0", "1" = "#ddcc77")

four_col <- c("0" = "#f0f0f0", "1" = "#117733")

binnary_colors <- c("0" = "#f0f0f0", "1" = "#7570b3")

character_presence_absence <- presence <- c("no" = "#f0f0f0", "yes" = "#7570b3")

genotype_cols <- c("4.1.1" = "#EECC66", "4.3.1.1" = "#EE99AA",
                   "4.3.1.1.EA1" = "#6699CC", "4.3.1.2" = "#997700",
                   "4.3.1.2.EA2" = "#994455", "4.3.1.2.EA3" = "#004488")

plasmid_type_cols <- c("Col type" = "#EECC66", "IncF type" = "#EE99AA",
                       "IncH type" = "#6699CC", "IncR type" = "#997700",
                       "IncX type" = "#004488")

replicon_cols <- c("Col(pHAD28)" = "#71bc78", "Col440II" = "#93dfb8",
                   "Col440I" = "#45cea2", "ColRNAI" = "#17806d",
                   "FII(pBK30683)" ="#18a7b5",  "IncFIA(HI1)" = "#ace5ee",
                   "IncFIB(K)(pCAV1099-114)" ="#0d98ba",
                   "IncFIB(K)" = "#9aceeb", "IncFIB(pNDM-Mar)" = "#1164b4",
                   "IncFIB(pQil)" = "#EE99AA", "IncFII(K)" = "#994455",
                   "IncFII(pCRY)" = "#EECC66", "IncFII(pKP91)" = "#997700",
                   "IncFII(pKPX1)" = "#1a4876", "IncHI1B(pNDM-MAR)" = "#a2add0",
                   "IncR" = "#324ab2", "IncX1" = "#7366bd", "Other" = "#252525")


amr_class_cols <- c("Aminoglycosides" = "#CC6677", "Other" = "#252525",
                    "Beta-lactam" = "#332288", "Fosfomycin" = "#ddcc77",
                    "Fluoroquinolones" = "#117733", "Phenicols" = "#88ccee",
                    "Sulfonamides" = "#882255", "Tetracyclines" = "#44aa99",
                    "Trimethoprim" = "#999933")

beta_lactam_cols <- c("blaSHV" = "#848096", "blaOXA|TEM|LAP"  = "#6c658f",
                      "CTX-M-15" = "#4e428a", "NDM-1" = "#332288")

levels(as_factor(amr_4$beta_lactam_grp))
yersinabactin_st_cols <-
  c("145-2LV" = "#EECC66", "151" = "#EE99AA", "277" = "#6699CC",
    "294" = "#997700", "327" = "#994455", "327-1LV" = "#004488", "78" = "#DDDDDD")

# rpinecone output
lineage_col <- c("1" = "#EE8866")


sub_lineage_col <- c("1" = "#CC6677",
              "10" = "#332288",
              "11" = "#ddcc77",
              "12" = "#117733",
              "15" = "#88ccee",
              "2" = "#882255",
              "3" = "#44aa99",
              "5" = "#999933")


ont_col <- c("yes" = "black")
# saving directory
outdir <- "./results/plots/r/"
```


# load data 
```{r load_data, message=FALSE}

# the Panaroo gene presence-absence matrix
data <- read.table("/Users/zuza/repos/redesigned-lamp/results/panaroo/gene_presence_absence.Rtab",
                   header = TRUE)

# load the genes inserted into outbreak isolates
load(file = "data/r/outbreakCDSs.RData")

# load core gene tree from panaroo
core_gene_tree <-
  read.tree("results/core_gene_tree/core_tree.treefile")


# load the snp based tree
snp_tree <-
  ape::read.tree("./results/iqtree/phylo.aln.treefile")

#load tree produced from gubbins
gubbins_tree <-
  ape::read.tree("./results/gubbins/ST39.final_tree.tre")

# load the manually generated gubbins tree
manual_tree <-
  ape::read.tree("./results/snippy/iqtree/manual.filtered_polymorphic_sites.fasta.treefile")

# read metadata from the Genome paper
metadata <- 
    read_csv("docs/kleb_metadata.csv",
           show_col_types = FALSE) %>%
  clean_names() %>%
  filter(main_st == "ST39")

tip_metadata <-
  metadata %>%
  select(reference, strain, sample_acc, lane_acc, supplier_name,
         study_accession, supplier_sample_name, date, neonate, major_wards)


load(file = "data/r/rPinecone_lineages.RData")

ariba <-
  read_csv("./docs/clean_ariba_amr.csv") %>%
  remove_constant()

argannot <-
  read_table("./docs/cleaned_argannot.txt")

plasmidfinder <-
  read_tsv("./docs/plamidfinder_summary.txt")

colnames(plasmidfinder) <- str_replace_all(colnames(plasmidfinder), "_1", "")

head(plasmidfinder)

# kleborate results
kleborate <-
  read_tsv("./results/kleborate/kleborate_results.txt")

# load panaroo summary file
pangenome_summary <-
  read_csv("./docs/pangenome_summary.csv",
           col_names = c("Gene Type", "Gene Count"))

pangenome_distribution <-
  read_tsv("./results/panaroo/qc/ngenes.txt")

```

# plot the isolate distribution
```{r}

    metadata %>% mutate(year = factor(date)) %>%
    mutate(ward = factor(major_wards)) %>%
    ggplot(aes(x = year, fill = ward)) +
    geom_bar() +
    scale_fill_manual(values = ward_col) +
  labs(y = "Number of Genomes",
       x = "Year of Isolation",
       fill = "Ward") +
  theme(axis.text = element_text(size = 13,
                                 face = "bold"),
        axis.text.x = element_text(angle = 30,
                                   vjust = 0.5),
        axis.title = element_text(face = "bold",
                                  size = "15"),
        legend.title = element_text(size = 15,
                                    face = "bold"))

```


# Plot the pangenome summary
```{r}
summary <-
  pangenome_summary %>%
  clean_names() %>%
  mutate(Group = ifelse(gene_type == "Total genes", "Total Genes", "Other")) %>%
  ggplot(aes(x = reorder(gene_type, - gene_count), y = gene_count, fill = Group)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = gene_count), vjust = -0.5) +
  scale_fill_manual(values = c("Total Genes" = "#332288")) +
  labs(
    x = "Gene Type",
    y = "Gene Count"
  ) +
  guides(fill = "none") +
  vexpand(0.01)

distribution <-
  pangenome_distribution %>%
  pivot_longer(cols = "no_genes") %>%
  ggplot(aes(x = name, y = value, fill = name)) +
  geom_violin(show.legend = FALSE) +
  theme_minimal() +
  labs(
    x = "",
    y = "Number of Genes"
  ) +
  scale_fill_viridis_d()
distribution + theme(axis.text.x = element_blank())
```


# Prepare metadata for annotating heatmap

## ONT data
```{r}
ids_with_ont <-
  c("ERR12059348", "ERR12058301", "ERR12059076", "ERR12058306", "ERR12058885")
```


## prepare lineage annotation
```{r , eval=FALSE}

#create DNAbin from alignment
#msa <- "./results/snp_sites/phylo.fasta"
msa <- "./results/snippy/20240925_2/manual.filtered_polymorphic_sites.fasta"

my_alignment <-
  fasta2DNAbin(msa, chunkSize = 10, snpOnly = FALSE)

# create a snp scaled tree
snp_scaled_tree <- rPinecone::ace_scale_phylo(manual_tree, my_alignment)

# perform lineage asignment
results <- pinecone(snp_scaled_tree, 5, 5)

saveRDS(results, "data/r/rPinecone_lineages.RData")

lineage_1 <- results$table %>% as.data.frame()

rownames(lineage_1) <- lineage_1$Taxa
lineage_1

lineage_1 <- lineage_1 %>% select(!c(Taxa))

lineage_1$sub_lineage <-
  case_when(grepl("singleton", lineage_1$`Sub-lineage`) ~ "singleton",
                                   TRUE ~ lineage_1$`Sub-lineage`)

major_lineages <-
  lineage_1 %>%
  select(`Major.Sub-lineage`) %>%
  mutate(across(where(is.character), as.factor))


sub_lineages <-
  lineage_1 %>%
  select(sub_lineage) %>%
  mutate(across(where(is.character), as.factor))

sub_lineages$sub_lineage 

outbreak_clade <- sub_lineages %>% filter(sub_lineage == 8)
itol_labels_template(results)
results$table
str(sub_lineages)
summary(sub_lineages$sub_lineage)
```

## prepare date annotation

```{r common_annotation_files}
# load date information from metadata file
date_data <- 
  tip_metadata %>%
  select(c(lane_acc, date, major_wards)) %>%
  filter(lane_acc %in% core_gene_tree$tip.label)


date_data_1 <-
  date_data %>%
  mutate(ont =
           case_when(date_data$lane_acc %in% ids_with_ont ~ "yes",
                         TRUE ~ ""),
         year =
           case_when(date_data$date == 2016 ~ "2016",
            date_data$date == 2017 ~ "2017",
            date_data$date == 2018 ~ "2018",
            date_data$date == 2019 ~ "2019",
            TRUE ~ "Other"),
         wards =
           case_when(date_data$major_wards ==
              "CHATINKHA NURSERY" ~ "CHATINKHA NURSERY",
            TRUE ~ "Other wards"))


# again convert to a data.frame object
meta1 <- as.data.frame(date_data_1)

# assign rows
rownames(meta1) <- meta1$lane_acc

meta2 <- meta1 %>% select(year)

meta3 <- meta2 %>%
  mutate(year = as.factor(year))

# a column indicating isolates with ont data
ont_df <- meta1 %>% select(ont)

```

## Prepare kleborate output
```{r}
unique_kleborate <-
  kleborate %>% remove_constant()


df1 <-
  unique_kleborate %>%
  select(strain, YbST, ybtS, ybtX, ybtQ, ybtA, irp2, irp1, ybtU, ybtT, ybtE, fyuA)

df1<- 
  df1 %>%
  mutate(across(where(is.character), as.factor))

df1[2:12] %>% summary()
df2 <- as.data.frame(df1)
rownames(df2) <- df2$strain

df3 <-
  df2 %>%
  select(!c(strain))

ybst <-
  df3 %>%
  select(YbST)


yb_genotype <-
  df3 %>%
  select(!c(YbST))



colnames(unique_kleborate)
unique_kleborate %>%
  mutate(across(where(is.double), as.factor)) %>%
  summary

unique_kleborate %>% group_by(iroD) %>% summary()

```

## annotation for the core_gene tree
the dataframe from panaroo represents binary values as characters. there is a need
to convert them to numeric values then to binnary for optimal annotation
```{r create_heatmap_columns}

all_genes <- c(cds$gene, "gpFI", "xerC_1", "traR", "pknD", "pknH",
                     "dnaB_1", "capE", "cdnE", "capV", "dncV")

# backbone genes for the regions
backbone_genes <-
  c("evgA", "hdeB_2", "mug", "bepF", "rob_1", "pgrR_2", "smpB", "ratA",
    "dmlR", "ycaC", "dsbA", "phoH", "efeB", "galF", "gnd", "ugd", "wbgU",
    "tagG", "hisI", "hisF", "hisA", "hisB", "hisC_1", "hisD", "hisG", "yeeZ" ,
    "yhaJ_2", "plaP", "fusA_2", "sbcB", "dacD",
    "sbmC", "yeeA", "cobS", "cobT_2", "ahcY", "dmlR_8", "abaF_1",
    "yeaR_1", "gcvA_2", "cpo", "nimR_2", "aaeB_2", "aaeA_2", "fabG_3",
    "sotB_1", "intA_2", "yeeO", "amn", "pmrD", "menH_3", "acuI", "acuR")

# function to create gene annotation files

make_heatmap <-
  function(genes) {
  df1 <-
    data %>% 
    filter(Gene %in% genes$gene) #%>%
    #filter(Gene %notin% backbone_genes)
  
  #transpose the filtered matrix
  transposed_df <- t(df1)
  
  colnames(transposed_df) <-transposed_df[1,]
  
  # filter out the gene column
  restructured_df <- transposed_df[rownames(transposed_df) %notin% c("Gene"), ]
  
  # convert to a data.frame object because tibble handles rownames in a wierd way!
  segment_genes <- as.data.frame(restructured_df)
  
  genes <-
  segment_genes %>%
  mutate(across(where(is.character), as.factor))
  
  return(genes)
  }

# dataframes for heatmaps of the regions identified using graphia
one <- make_heatmap(cds %>% filter(region == "One"))

two <- make_heatmap(cds %>% filter(region == "Two"))

three <- make_heatmap(cds %>% filter(region == "Three"))

four <- make_heatmap(cds %>% filter(region == "Four"))

```

## annotations for the snp-based tree
```{r snp_based_tree_annotations}
amr_df <-
  argannot %>%
  as.data.frame() 

rownames(amr_df) <-
  amr_df$"#FILE"

amr_df_num <- select(amr_df, !c("#FILE", "NUM_FOUND",
                                "Penicillin_Binding_Protein_Ecoli",
                                "FosA6", "OqxA", "OqxBgb", "ampH"))

colnames(amr_df_num)
# replace all numeric to binary by matching the missing values
for (i in colnames(amr_df_num)) {
  amr_df_num[ ,i] <- if_else(amr_df_num[ ,i] %in% ".", 0, 1)
}

amr_df1 <-
  amr_df_num %>%
  mutate(across(where(is.numeric), as.factor))

head(amr_df1)

# plasmid data
plasmids <-
  plasmidfinder %>%
  as.data.frame()

# this column has extra characters which mess with parsing to subset the values
plasmids$`Col(pHAD28)` <-
  substr(plasmids$`Col(pHAD28)`, 1,4)

rownames(plasmids) <-
  plasmids$`#FILE`

plasmid_data <-
  select(plasmids, !c(`#FILE`, NUM_FOUND))

# replace all numeric to binary by matching the missing values
for (i in colnames(plasmid_data)) {
  plasmid_data[ ,i] <- if_else(plasmid_data[ ,i] %in% ".", 0, 1)
}


plasmid_data_1 <-
  plasmid_data %>%
  mutate(across(where(is.numeric), as.factor))

```
# create summary plots
## summary plots of plasmid data
```{r}
y <- plasmid_data

y$genome <- rownames(y)

y$raw_date <- case_when(y$genome %in% tip_metadata$lane_acc ~ tip_metadata$date)
y$date <- case_when(y$genome %in% date_data$lane_acc ~ date_data$year,
                    TRUE ~ "Other")

# calculate proportions for the results sections
plasmid_columns <- colnames(y[1:17]) %>% toString()

# set outbreak clade to use for comparison of frequencies
outbreak_clade$genome <- rownames(outbreak_clade) 

avrg #<-
  y %>% mutate(clade = case_when(genome %in% outbreak_clade$genome ~ "Outbreak",
                               TRUE ~ "Other")) %>%
  rowwise() %>%
    mutate(average = sum(c(`Col(pHAD28)`, Col440II, Col440I, ColRNAI, `FII(pBK30683)`,
                           `IncFIA(HI1)`, `IncFIB(K)(pCAV1099-114)`,
                           `IncFIB(K)`, `IncFIB(pNDM-Mar)`, `IncFIB(pQil)`,
                           `IncFII(K)`, `IncFII(pCRY)`, `IncFII(pKP91)`,
                           `IncFII(pKPX1)`, `IncHI1B(pNDM-MAR)`,
                           IncR, IncX1))) %>%
  select(clade, average) %>%
    filter(clade == "Other") %>%
    summary()


freq <- y %>% mutate(clade = case_when(genome %in% outbreak_clade$genome ~ "Outbreak",
                               TRUE ~ "Other")) %>%
  group_by(clade) %>%
  select(Col440I, `IncFIB(K)`, ColRNAI, clade) #%>% 
  summarise_all(sum) %>% pivot_longer(cols = Col440I:ColRNAI) %>%
  ggplot(mapping = aes(x = clade, y = value, fill = name)) +
  geom_col(position = "fill")

 
y %>% group_by(date) %>% select(Col440I, `IncFIB(K)`, ColRNAI, date) %>%
  summarise_all(sum) %>%
  pivot_longer(cols = Col440I:ColRNAI) %>%
  ggplot(mapping = aes(x = date, y = value, fill = name)) +
  geom_col(position = "fill") +
  theme_light() +
  labs(y = "Proportion of Plasmid Replicon",
       x = "Date",
       fill = "Replicon Type")

y
# calculate proportions for plasmids
y[1:17] %>%
  as_tibble() %>%
  summarise_all(sum) %>%
  pivot_longer(cols = `Col(pHAD28)`:IncX1) %>%
  mutate(prop = value/sum(value)*100)


piv1 <- y %>% pivot_longer(cols = `Col(pHAD28)`:IncX1,
                   names_to = "replicon",
                   values_to = "presence")


piv1

piv1 %>% ggplot() + geom_col(aes(date, presence, fill = replicon))
#y <- y %>% select(!c(genome))

# create a column plot with proportions vs date
plasmid_prps <- piv1 %>% ggplot(mapping = aes(x = date, y = presence, fill = replicon)) +
  geom_col(position = "fill") +
  theme_light() +
  labs(y = "Proportion of Plasmid Replicon",
       x = "Date",
       fill = "Replicon Type") +
  scale_fill_manual(values = replicon_cols) +
  theme(axis.text = element_text(size = 13,
                                 face = "bold"),
        axis.text.x = element_text(angle = 30,
                                   vjust = 0.5),
        axis.title = element_text(face = "bold",
                                  size = "15"),
        legend.title = element_text(size = 15,
                                    face = "bold"))

  
```

## summary plots for the AMR data
```{r eval=FALSE}

amr_1 <- amr_df1

# create a column with the accession numbers
amr_1$genome <- rownames(amr_1)

# remove the assemblies from the ONT data
amr_2 <-
  amr_1 %>%
  filter(genome != c("CAA8N9")) %>%
  filter(genome != c("CAAUE8"))


# import the date information
amr_2$date <-
  case_when(amr_2$genome %in% date_data_1$lane_acc ~ date_data_1$year,
            TRUE ~ "Other")



amr_2

# calculate proportions for plasmids

num_amr <-
  amr_2 %>% mutate(clade = case_when(genome %in% outbreak_clade_ids ~ "Outbreak",
                               TRUE ~ "Other")) %>%
  rowwise() %>%
  mutate(across(where(is.factor), as.character)) %>%
  mutate_at(vars(colnames(amr_2[1:42])), as.numeric) 

num_amr %>%
  rowwise() %>%
  mutate(total = sum(c(`aac6-Ib`, dfrA14))) %>% print(n = ,1000)


num_amr %>%
  rowwise() %>%
  mutate(total = sum(c(`aac3-IIa`, `aac3-IId`, `aac6-Ib`, `aadA1-pm`,
                       aadA16, aadA2, `aph3-Ia`, aphA6,  strA,  strB,
                       `blaCTX-M-15`, `blaLAP-2`, `blaNDM-1`, `blaOXA-1`,
                       `blaOXA-10`, `blaOXA-9`, `blaSHV-1`, `blaSHV-187`,
                       `blaTEM-105`, `qnr-S1`, qnrB1,
                       qnrB17, `mph(A)`, catA1, catA2, catB4, cmlA5, arr2,
                       arr3, sul1, sul2, tetA, tetD, tetR, dfrA12, dfrA14,
                       dfrA15b, dfrA27, dfrA30, dhfr7)))  %>%
  select(clade, total) %>%
    filter(clade == "Other") %>%
    summary()

num_amr 

rfc_amr <-
  num_amr[1:42] %>%
  as_tibble() %>%
  summarise_all(sum) %>%
  pivot_longer(cols = `aac3-IIa`:dhfr7) %>%
  mutate(prop = value/135*100)

View(rfc_amr)

order(rfc_amr$value)
# the presence/absebce information is in factor format
# direct conversion into numeric format is bringing unusal formating
# convert them to characterfirst then into numeric
amr_2_fct <-
  amr_2 %>%
  mutate(across(where(is.factor), as.character))

# convert into long formart
amr_3 <- 
  amr_2_fct %>% pivot_longer(cols = `aac3-IIa`:dhfr7,
                   names_to = "gene",
                   values_to = "presence")

# classify the antibiotic genes acording to drug class
amr_4 <- amr_3 %>%
  mutate(class = case_when(grepl("aa|aph|str", amr_3$gene) ~ "Aminoglycosides",
                           grepl("bla", amr_3$gene) ~ "Beta-lactam",
                           grepl("tet", amr_3$gene) ~ "Tetracyclines",
                           grepl("qnr", amr_3$gene) ~ "Fluoroquinolones",
                           grepl("cat|cmlA5", amr_3$gene) ~ "Phenicols",
                           grepl("sul", amr_3$gene) ~ "Sulfonamides",
                           grepl("dfr|dhf", amr_3$gene) ~ "Trimethoprim",
                           grepl("Fos", amr_3$gene) ~ "Fosfomycin",
                           TRUE ~ "Other"),
         beta_lactam_grp = case_when(grepl("SHV", amr_3$gene) ~ "blaSHV",
                                     grepl("OXA|TEM|LAP", amr_3$gene) ~ "blaOXA|TEM|LAP",
                                     grepl("CTX", amr_3$gene) ~ "CTX-M-15",
                                     grepl("NDM", amr_3$gene) ~ "NDM-1",
                                     TRUE ~ "Other"),
         beta_lactam_fct = factor(beta_lactam_grp,
                     levels = c("blaSHV", "blaOXA|TEM|LAP", "CTX-M-15", "NDM-1", "Other"), ordered = T),
         num_presence = as.numeric(presence))


# plot the proportions of each gene class for every year  
all_amr <-
  amr_4 %>%
  ggplot(mapping = aes(x = date, y = num_presence, fill = class)) +
  geom_col(position = "fill") +
  theme_light() +
  labs(y = "Proportion of AMR gene class",
       x = "Date",
       fill = "AMR gene class") +
  scale_fill_manual(values = amr_class_cols) +
  theme(axis.text = element_text(size = 13,
                                 face = "bold"),
        axis.text.x = element_text(angle = 30,
                                   vjust = 0.5),
        axis.title = element_text(face = "bold",
                                  size = "15"),
        legend.title = element_text(size = 15,
                                    face = "bold"))


# subset the beta lactam genes
beta_lactamse <- 
  amr_4 %>%
  filter(class == "Beta-lactam") %>%
  ggplot(mapping = aes(x = date, y = num_presence, fill = beta_lactam_fct)) +
  geom_col(position = "fill") +
  theme_light() +
  labs(y = "Proportion of Beta-lactamase",
       x = "Date",
       fill = "Bet-lactamase gene") +
  scale_fill_manual(values = beta_lactam_cols) +
  theme(axis.text = element_text(size = 13,
                                 face = "bold"),
        axis.text.x = element_text(angle = 30,
                                   vjust = 0.5),
        axis.title = element_text(face = "bold",
                                  size = "15"),
        legend.title = element_text(size = 15,
                                    face = "bold"))

all_amr + beta_lactamse + plot_layout(guides = "collect")

```

# get outbreak clade isolate ids
```{r}
tree <- midpoint(manual_tree)

View(as_tibble(tree))
ggtree(tree) + geom_cladelab(node = 195, label = "Outbreak Isolates", barcolour = "red",
                             align = F) +
  geom_hilight(node=195, fill="#6699CC")

outbreak_clade_ids <- get_taxa_name(ggtree(tree), node = 195)

# display node labels
ggtree(tree) + geom_text(aes(label=node), hjust=-.3)
```



# map the three regions against the tree
```{r plot_core_gene_tree, message=FALSE}


text_size <- 2
y_offset <- 2.2

# using the tree generated after running iqtree on the gubbins recombination free alignment
tree <- midpoint(manual_tree) # from library phangorn

tree$tip.label <-
  tree$tip.label %>%
  str_replace_all( "Reference", "ERR12058301")

y <-
  tree %>%
  ggtree() %<+%
  date_data +
    geom_tippoint(aes(color=wards), size = 1.2) +
     scale_colour_manual(values = ward_col, na.translate = FALSE,
                        guide = guide_legend(order = 1)) +
  geom_treescale(#label = "point mutations",
                 y = 12,
                 x  = 0.0000006,
                 fontsize = 4,
                 linesize = 1)
y 

h1 <- 
  gheatmap(y, meta3,
               width = 0.02,
         colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = text_size) +
  scale_fill_manual(values = datecol,
                    na.translate = FALSE,
                        guide = guide_legend(order = 2)) +
  labs(fill = "Year")

h1

h2 <- h1 + new_scale_fill()

h3 <- 
  gheatmap(h2, one,
         offset = 0.000002,
         width = 0.17,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = text_size) +
  scale_fill_manual(values = one_col,
                    na.translate = FALSE,
                       labels = c("Absent", "Present"),
                    guide = guide_legend(order = 3)) +
  labs(fill = "Region One CDS") 

h3

h4 <- h3 + new_scale_fill()

h5 <- 
  gheatmap(h4, two,
         offset = 0.000009,
         width = 1.2,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = text_size) +
  scale_fill_manual(values = two_col,
                    na.translate = FALSE,
                       labels = c("Absent", "Present"),
                    guide = guide_legend(order = 4)) +
  labs(fill = "Region Two CDS") 

h5

h6 <- h5 + new_scale_fill()

h7 <- 
  gheatmap(h6, three,
         offset = 0.000055,
         width = 1,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = text_size) +
  scale_fill_manual(values = three_col,
                    na.translate = FALSE,
                       labels = c("Absent", "Present"),
                    guide = guide_legend(order = 5)) +
  labs(fill = "Region Three CDS") 

h7 +
  vexpand(0.02) + theme(
    legend.position = c(.05, 1),
    legend.justification = c("left", "top"),
    legend.box.just = "left",
    legend.margin = margin(6, 6, 6, 6)
    )


```


```{r write_plot-TO-file, eval=FALSE}
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "core_gene_with_genes.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

pdf(plot_name,
    width = 20,
    height = 15)

h7 +
  vexpand(0.02) + theme(
    legend.position = c(.05, .90),
    legend.justification = c("left", "top"),
    legend.box.just = "left",
    legend.margin = margin(6, 6, 6, 6),
    text = element_text(size = 30),
    axis.text = element_blank()
    )

dev.off() 
```


# simplified tree
```{r}
x = 1.5
y <-
  tree %>%
  ggtree() %<+%
  date_data_1 +
    geom_tippoint(aes(color=wards), size = 2) +
     scale_color_manual(values = ward_col ,
                        na.translate = FALSE,
                        guide = guide_legend(order = 1)) +
  geom_treescale(#label = "point mutations",
                 y = 12,
                 x  = 0.0000006,
                 fontsize = 4,
                 linesize = 1) +
    geom_tiplab(align = TRUE,
                size = 2,
                linesize = 0.2) +
  geom_hilight(node=195, fill="#6699CC", to.bottom = TRUE)
y 

ont_heatmap <- 
  gheatmap(y, ont_df,
               width = 0.01,
           offset = 0.000002 * x,
         colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = 3,
        hjust = 0.5,
        font.size = font_size) +
  scale_fill_manual(values = ont_col,
                    na.translate = FALSE,
                        guide = guide_legend(order = 1)) +
 labs(fill = "ONT Available")

ont_heatmap

ont_1 <- ont_heatmap + new_scale_fill()

h1 <- 
  gheatmap(ont_1, meta3,
               width = 0.01,
           offset = 0.0000024 * x,
         colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = 3,
        hjust = 0.5,
        font.size = font_size) +
  scale_fill_manual(values = datecol,
                    na.translate = FALSE,
                        guide = guide_legend(order = 2)) +
  labs(fill = "Year")

h1

h2 <- h1 + new_scale_fill()


sub_lineage <- 
  gheatmap(h2, sub_lineages,
         offset = 0.0000028 * x,
         width = 0.01,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = 4,
        hjust = 0.5,
        font.size = font_size) +
  scale_fill_manual(values = sub_lineage_col,
                       #labels = c("Absent", "Present"),
                    na.translate = FALSE,
                    guide = guide_legend(order = 3)) +
  labs(fill = "rPinecone Sub-lineages") 

sub_lineage

ybt <- sub_lineage + new_scale_fill()

h3_yb <- 
  gheatmap(ybt, ybst,
         offset = 0.0000032 * x,
         width = 0.01,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = 4,
        hjust = 0.5,
        font.size = font_size) +
  scale_fill_manual(values = yersinabactin_st_cols,
                    na.translate = FALSE,
                    guide = guide_legend(order = 4)) +
  labs(fill = "ybST") 

h3_yb

amr <- h3_yb + new_scale_fill()


h3_amr <-
  gheatmap(amr, amr_df1,
         offset = 0.0000040 * x,
         width = 0.4,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = 4,
        hjust = 0.5,
        font.size = font_size) +
      scale_fill_manual(values = binnary_colors,
                    na.translate = FALSE) +
  guides(fill = "none")

h3_amr

plasmid <- h3_amr + new_scale_fill()

h3_plasmid <- 
  gheatmap(plasmid, plasmid_data_1,
         offset = 0.000022,
         width = 0.2,
           colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = 4,
        hjust = 0.5,
        font.size = font_size) +
      scale_fill_manual(values = binnary_colors,
                    na.translate = FALSE,
                    labels = c("Absent", "Present"),
                    guide = guide_legend(order = 5)) +
  labs(fill = "AMR/plasmid") 

h3_plasmid + theme(
    legend.position = c(0.04, 1),
    legend.justification = c("left", "top"),
    legend.box.just = "left",
    legend.margin = margin(6, 6, 6, 6),
    text = element_text(size = 10)
    ) +
  vexpand(0.02)
```


# snp scalled tree
```{r}

tree <- midpoint(results$tree) # from library phangorn

tree %>%
  ggtree() %<+%
  date_data_1 +
    geom_tippoint(aes(color=wards), size = 1.2) +
     scale_colour_manual(values = ward_col, na.translate = FALSE,
                        guide = guide_legend(order = 1)) #+
  geom_treescale(label = "SNPs",
                 y = 20,
                 x  = 6,
                 fontsize = 4,
                 linesize = 1) +
   theme(
    legend.position = c(.06, 1),
    legend.justification = c("left", "top"),
    legend.box.just = "left",
    legend.margin = margin(6, 6, 6, 6),
    text = element_text(size = 10)
    )

```

```{r}
sessionInfo()
```

